tsdb/engine/tsm1/file_store.go

```diff
@@ -633,6 +629,43 @@ func (f *FileStore) BlockCount(path string, idx int) int {
 	return 0
 }
 
+// walkFiles calls fn for every files in filestore in parallel
+func (f *FileStore) walkFiles(fn func(f TSMFile) error) error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	// struct to hold the result of opening each reader in a goroutine
+	type res struct {
+		err error
+	}
+
+	resC := make(chan res)
+	var n int
+
+	for _, f := range f.files {
+		n++
+
+		go func(tsm TSMFile) {
+			if err := fn(tsm); err != nil {
+				resC <- res{err: fmt.Errorf("file %s: %s", tsm.Path(), err)}
+				return
+			}
+
+			resC <- res{}
+		}(f)
+	}
+
+	var err error
+	for i := 0; i < n; i++ {
+		res := <-resC
+		if res.err != nil {
+			err = res.err
```

Comment:
Do you specifically want the last error encountered here? If not (and you only care about _any_ error), you could drop the `var err error` variable, and just return once you hit `res.err != nil`.